timemachine/exchange/exchange_mover.py

```diff
@@ -87,108 +95,385 @@ def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff):
+        self.batch_log_weights = batch_log_weights
 
     def propose(self, x: CoordsVelBox) -> Tuple[CoordsVelBox, float]:
         coords = x.coords
         box = x.box
-        n_atoms = len(coords)
-        chosen_water = np.random.randint(self.num_waters)
-        chosen_water_atoms = self.water_idxs[chosen_water]
-
-        # compute delta_U of deletion
-        a_idxs = chosen_water_atoms
-        b_idxs = np.delete(np.arange(n_atoms), a_idxs)
+        log_weights_before = self.batch_log_weights(coords, box)
+        log_probs_before = log_weights_before - logsumexp(log_weights_before)
+        probs_before = np.exp(log_probs_before)
+        chosen_water = np.random.choice(np.arange(self.num_waters), p=probs_before)
+        chosen_water_atoms = self.water_idxs_np[chosen_water]
 
         # compute delta_U of insertion
         trial_chosen_coords = coords[chosen_water_atoms]
         trial_translation = np.diag(box) * np.random.rand(3)
         # tbd - what should we do  with velocities?
 
-        # this has a higher acceptance probability than if we allowed for rotations
-        # (probably because we have a much higher chance of a useless move)
-        moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
-        trial_coords = coords.copy()  # can optimize this later if needed
-        trial_coords[chosen_water_atoms] = moved_coords
+        moved_coords = randomly_rotate_and_translate(trial_chosen_coords, trial_translation)
+        # moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
+
+        # optimized version using double transposition
+        log_weights_after, trial_coords = self.batch_log_weights_incremental(
+            coords, box, chosen_water, moved_coords, log_weights_before
+        )
+
+        # reference version
+        # trial_coords = coords.copy()  # can optimize this later if needed
+        # trial_coords[chosen_water_atoms] = moved_coords
+        # log_weights_after = self.batch_log_weights(trial_coords, box)
+
+        log_p_accept = min(0, logsumexp(log_weights_before) - logsumexp(log_weights_after))
+        new_state = CoordsVelBox(trial_coords, x.velocities, x.box)
+
+        return new_state, log_p_accept
+
 
-        delta_U_insert = self.U_fn(trial_coords, box, a_idxs, b_idxs)
-        # If our system is in a clash free state, a deletion move has delta_Us typically
-        # on the order of 35kTs. However, if we start in a clashy state, we can no longer
-        # guarantee this. So it's safer to disable this micro-optimization for now.
-        delta_U_delete = -self.U_fn(coords, box, a_idxs, b_idxs)
-        delta_U_total = delta_U_delete + delta_U_insert
+# numpy version of delta_r
+def delta_r_np(ri, rj, box):
+    diff = ri - rj  # this can be either N,N,3 or B,3
+    if box is not None:
+        box_diag = np.diag(box)
+        diff -= box_diag * np.floor(diff / box_diag + 0.5)
+    return diff
+
+
+def v1_insertion(radius, center, box):
```

Comment:
nit: name, perhaps `vol1_insertion`, `vol2_insertion` or `sample_within_sphere`, `sample_outside_sphere`?

(initially misread as "version 1 insertion", "version 2 insertion")

Reply:
willfix

Reply:
renamed to `inner_insertion` and `outer_insertion`

Reply:
(also renamed `v1_mols` to `inner_mols` and `v2_mols` to `outer_mols`)